lib/IDE/Refactoring.cpp

```diff
@@ -6473,18 +6599,64 @@ class AsyncConverter : private SourceEntityWalker {
     if (!Args.empty()) {
       if (AddedReturnOrThrow)
         OS << " ";
-      if (Args.size() > 1)
-        OS << tok::l_paren;
-      for (size_t I = 0, E = Args.size(); I < E; ++I) {
-        if (I > 0)
-          OS << tok::comma << " ";
+      unsigned I = 0;
+      addTupleOf(Args, OS, [&](Expr *Elt) {
         // Can't just add the range as we need to perform replacements
-        convertNode(Args[I], /*StartOverride=*/CE->getArgumentLabelLoc(I),
+        convertNode(Elt, /*StartOverride=*/CE->getArgumentLabelLoc(I),
                     /*ConvertCalls=*/false);
+        I++;
+      });
+    }
+  }
+
+  /// Assuming that \p CE is a call to \c TopHandler's completion handler and
+  /// that the current scope is wrapped in a continuation, replace it with a
+  /// call to the continuation.
+  void addHandlerCallToContinuation(const CallExpr *CE) {
+    assert(TopHandler.getAsHandlerCall(const_cast<CallExpr *>(CE)) == CE &&
+           "addHandlerCallToContinuation must be used with a call to the "
+           "TopHandler's completion handler");
+    assert(Scopes.back().isWrappedInContination());
+
+    ArrayRef<Expr *> Args;
+    StringRef ResumeArgumentLabel;
+    switch (TopHandler.getHandlerType()) {
+    case HandlerType::PARAMS: {
+      auto Exprs = TopHandler.extractResultArgs(CE);
+      Args = Exprs.args();
+      if (!Exprs.isError()) {
+        ResumeArgumentLabel = "returning";
+      } else {
+        ResumeArgumentLabel = "throwing";
       }
-      if (Args.size() > 1)
-        OS << tok::r_paren;
+      break;
+    }
+    case HandlerType::RESULT: {
+      Args = callArgs(CE).ref();
+      ResumeArgumentLabel = "with";
+      break;
+    }
+    case HandlerType::INVALID:
+      llvm_unreachable("Invalid top handler");
     }
+
+    Identifier ContName = Scopes.back().ContinuationName;
+    OS << ContName << tok::period << "resume" << tok::l_paren
+       << ResumeArgumentLabel << tok::colon << ' ';
+
+    if (Args.empty()) {
+      // Emit '()' to pass a 'Void' tuple to the continuation.
+      OS << tok::l_paren << tok::r_paren;
```

Comment:
I believe `addTupleOf` will do this anyway for an empty array

Reply:
Indeed. That’s simpler. I updated it.